### DISCUSSION TITLE
[sdk-js]: Proposal - Exception Handling & Negative Testing

### DISCUSSION THREAD

**Discussion by eliashussary** [directus/directus#3482] [Feature Requests]
## Exception Handling
`axios` throws an exception for any status outside of the range of `2xx` - in theory this makes sense, but in practice it's may not be ideal.

when using the directus sdk, `4xx` errors thrown back by the API are sort of hidden; the consumer needs to explicitly catch the exception and handle the `axios` error.

IMHO, an exception should be reserved for `5xx`  or when something is *really* wrong.  (more aligned with standard `fetch`)

for me to handle `4xx` errors in the current state requires additional boilerplate;

```ts
try{
  const res = await directus.items("content_page").read(query);
  // do something with response
}catch(err){
   if(err.status < 500){
     const directusErrors = err.response.data.errors
     // do soemthing with exception
   }
}
```

Example of error:
![image](https://user-images.githubusercontent.com/7349258/102721063-590e4f80-42c6-11eb-9bd9-7effc6aaac05.png)

i feel a more ideal state would be to simply return the response, without exception, and reserve exceptions for an event which is more critical;

```ts
const res = await directus.items("content_page").read(query);
if(res.errors){
   // do something with errors
}
```

### Negative tests
I did not see negative tests for these scenarios at all, so i've had a hard time understanding what the intent was on handling these errors.

happy to contribute and open a PR, but wanted to see what the plan was or is on this.

**Comment by rijkvanzanten** [comment]
@eliashussary Agreed! Sounds like a solid plan. Feel free to open a PR for that 👍

**Comment by apstanisic** [comment]
What do you think of having something like `Option` or `Either` for result, so user can access data only after checking that there is no error? 
Here's a cool library [https://gcanti.github.io/fp-ts/modules/Either.ts.html](https://gcanti.github.io/fp-ts/modules/Either.ts.html)
And here's a snippet from one of my projects that can be ported.

```ts

export type Either<T, E> = Result<T, undefined> | Result<undefined, E>;

/**
 * Result from functions with data, error properties
 * If data exist, there is no error, and reverse
 */
export class Result<Data = any, Error = any> {
  /**
   * Data from result
   */
  data: Data;
  /**
   * Error from result
   */
  error: Error;

  /**
   * Create new result
   * Do not invoke constructor directly, use factory methods
   */
  private constructor({ data, error }: { data: Data; error: Error }) {
    if (data === undefined && error === undefined) throw 'Both values can not be defined';
    if (data !== undefined && error !== undefined) throw 'Both values can not be undefined';

    this.data = data;
    this.error = error;
  }

  /** Call this on success */
  static Success<T>(data: T): Result<T, undefined> {
    return new Result<T, undefined>({ data, error: undefined });
  }

  /** Call this on fail */
  static Fail<E>(error: E): Result<undefined, E> {
    return new Result<undefined, E>({ error, data: undefined });
  }
}
```

**Reply by rijkvanzanten** [reply]
Can you give an example of how that'd look when using the SDK?

**Reply by eliashussary** [reply]
@apstanisic looks like you're referencing what looks to me like the `lodash` of types :) 

i can incorporate types to handle the possibility of an error... no problem; not sure it's worth adding in a typings lib just for `Either` we can achieve the same with generics out of box imho

**Reply by apstanisic** [reply]
Yeah, it's probably good idea to not add another dependency just for this.
Here's an example. I just used snippet above.
Before checking if data or error is undefined, both value will be nullable, after checking that error is undefined, data will become non nullable, and vice versa.
 

[https://codesandbox.io/s/gallant-chatelet-jmy42?file=/src/index.ts](https://codesandbox.io/s/gallant-chatelet-jmy42?file=/src/index.ts)

Here's code from codesandbox:
```ts
import { Result } from "./result";
async function getData(id = 1) {
  if (id !== 1) {
    return Result.Fail({ error: "not 1" });
  } else {
    return Result.Success({ id: 1 });
  }
}

async function main() {
  const result = await getData();

  console.log(result.data.id);
  console.log(result.error.error);
  if (!result.error) {
    console.log(result.data.id);
  }
  if (!result.data) {
    console.log(result.error.error);
  }
}

```

**Reply by rijkvanzanten** [reply]
Ohh I see what you're going for. I'd rather stick to Promises with resolve/reject (eg async await), so we have a more "standard" async usage flow 👍

**Reply by apstanisic** [reply]
Sure! 🙂

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
